docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
+
+# Objects
+
+Each token can contain an arbitrary amount of keys and other object. (Note:
+other objects are not implemented yet). The keys are TPM-based keys, i.e. they
+are only known to the TPM and also the key operations are executed inside the
+TPM. The other objects are encrypted using hybrid encryption with the outer
+key being sealed to the TPM.
+
+Note: Since the TPM's object are stored (encrypted) external to the TPM on the
+HDD, it is not possible to delete them, if your attacker model allows access to
+the HDD.
+
+# Key Authorizations
+
+For each token, two TPM seal blobs exist one for the SO and one for the User.
```

Comment:
Why not use PIN Fail NV indices, and key policies that reference them?

Also, what's the point of the SO PIN here? Seems redundant when there's already owner authorization over the TPM.

Reply:
I wanted to avoid using any additional NV and use transient objects as much as possible. There also exists a goal to support password based objects, so existing objects can be migrated if the attributes allow for it and objects can be easily migrated out if needed. Most objects currently in use just have a password. With that said, the SOPIN allows the SO to reset the user password.

You're not guaranteed that you have owner authorization over the TPM. You have access to a persistent primary object, which could be under any hierarchy. Currently the primary object is rooted in the owner hierarchy, but you never grant owner hierarchy permissions to tokens/so. However, i'd like the hierarchy to be configurable at setup time.

Reply:
I see. I think maybe it would be helpful if the document somewhere specified the use cases and constraints it's targeting.

E.g., I'm interested in using tpm2-pkcs11 on my own personal devices, where I'm the TPM2 owner, and where the entire reason I'm using it is to be able to prevent PIN brute forcing and key ex-filtration. I understand TPM2 supports more complicated ownership models for enterprises, but I'm not personally familiar with them, and it's not been a priority for me to learn them yet.

I also expect that if I'm concerned that my PIN may have been compromised, that I can just change it and feel confident the old PINs will no longer work. My understanding of TPM2 is this sort of revocation can't happen without involving NV. (Maybe clock-based policies could provide some safety guarantee that old PINs eventually expire? But you'd have to re-seal the auth value periodically or risk getting locked out.)

I can appreciate that there's limited NV space on a TPM2, and so it's probably best to minimize how many indices are used anyway (e.g., recommending to group multiple keys onto a single "token" so they can share a single PIN / NV Index). But I think the default should be to use NV space; and for users who can't or don't want to use NV, to support a separate less secure mode that uses only transient objects (and has to be explicitly enabled, so they're aware of the reduced security).

Reply:
> I see. I think maybe it would be helpful if the document somewhere specified the use cases and constraints it's targeting.

My goal with the "document architecture doc" was to do that. Andreas was nice enough to take a stab at it so he understood the constraints as hes working on an internals change as we speak.

> 
> E.g., I'm interested in using tpm2-pkcs11 on my own personal devices, where I'm the TPM2 owner, and where the entire reason I'm using it is to be able to prevent PIN brute forcing and key ex-filtration. I understand TPM2 supports more complicated ownership models for enterprises, but I'm not personally familiar with them, and it's not been a priority for me to learn them yet.

This is unfortunately where technology meets people and weirdness ensues.

> 
> I also expect that if I'm concerned that my PIN may have been compromised, that I can just change it and feel confident the old PINs will no longer work. My understanding of TPM2 is this sort of revocation can't happen without involving NV. (Maybe clock-based policies could provide some safety guarantee that old PINs eventually expire? But you'd have to re-seal the auth value periodically or risk getting locked out.)
> 
> I can appreciate that there's limited NV space on a TPM2, and so it's probably best to minimize how many indices are used anyway (e.g., recommending to group multiple keys onto a single "token" so they can share a single PIN / NV Index). But I think the default should be to use NV space; and for users who can't or don't want to use NV, to support a separate less secure mode that uses only transient objects (and has to be explicitly enabled, so they're aware of the reduced security).

We're actually working up a new authmodel design that policy based and supports wrapping auth values too. We plan on writing that up and sharing it next week for you and others to comment on so we can incorporate it with the internals change. @idesai in the new model, on the policy side, if the password changes does it invalidate old pins?